networking/core/src/worker.rs

```diff
@@ -636,29 +661,44 @@ where
                 },
             }
         } else {
-            match self
-                .gossip_message_codec
+            // the incoming gossip message is a transaction
+            if let Ok((length, msg)) = self
+                .transaction_gossip_message_codec
```

Comment:
Maybe we should inspect the topic prefix rather than trying to decode and failing - as this might be slow or even potentially erroneously succeed for the wrong message type (because protobuf works on optional fields and ignores unknown fields).

We could "register" channels for each topic or topic prefix.

e.g. 

```rust
networking.gossipsub_subscribe("consensus-0-31", codec);
```

or maybe register a channel for sending/receiving messages on the topic

```rust
let channel = MessageChannel::<DanMessage>::new();
networking.gossipsub_subscribe("consensus-0-31", channel);
```

Channel could be something like this

```rust
pub fn channel<T>(capacity: usize) -> (MessageChannel<T>, MessageChannel<T>) {
    let (ltr_tx, ltr_rx) = mpsc::channel(capacity);
    let (rtl_tx, rtl_rx) = mpsc::channel(capacity);
    (
        MessageChannel {
            sender: ltr_tx,
            receiver: rtl_rx,
        },
        MessageChannel {
            sender: rtl_tx,
            receiver: ltr_rx,
        },
    )
}

#[derive(Debug)]
pub struct MessageChannel<T> {
    sender: mpsc::Sender<T>,
    receiver: mpsc::Receiver<T>,
}

impl<T> MessageChannel<T> {
    pub async fn recv(&mut self) -> Option<T> {
        self.receiver.recv().await
    }

    pub fn try_recv(&mut self) -> Result<T, TryRecvError> {
        self.receiver.try_recv()
    }

    pub async fn send(&self, value: T) -> Result<(), SendError<T>> {
        self.sender.send(value).await
    }

    pub fn try_send(&self, value: T) -> Result<(), TrySendError<T>> {
        self.sender.try_send(value)
    }
}
```

Reply:
What might make it easier is if you do the en/decoding in the subscription handle and have the networking send the raw bytes across.

```rust
// Subscribe - that sets up a channel for sending/receiving
let (sender, receiver) = networking.gossipsub_subscribe("consensus-0-31").await?;
let msg = receiver.next_message().await?; 
sender.send(msg).await?;


#[derive(Debug, Clone)]
struct GossipSender<T> {
  sender: mpsc::UnboundedSender<Vec<u8>>,
  codec: ProstCodec<T>,
}

impl<T: prost::Message> GossipSubscription<T> {
   pub async fn send(&self, msg: T) -> Result<(), MessageError> {
        let len = msg.encoded_len();

       let mut buf = Vec::with_capacity(len);
        self.codec
              .encode_to(&mut buf, message)
              .await?;
        self.sender.send(buf)?;
       Ok(())
  }
}

#[derive(Debug)]
struct GossipReceiver<T> {
  receiver: mpsc::Receiver<gossipsub::Message>,
  codec: ProstCodec<T>,
}

impl<T: prost::Message> GossipSubscription<T> {
   pub async fn next_message(&self) -> Result<Option<(usize, T)>, MessageError> {
          let Some(raw_msg) =  self.receiver.recv().await? else {
            return Ok(None);
         };

        let (len, msg) = self.codec.decode_from(&mut raw_msg.data.as_slice()).await?;
       Ok((len, msg))
   }
}

```